### PR TITLE
Added MySql und MariaDB compatibility check to the setup.

### PR DESCRIPTION
**Pull request by guelzow**
Hi everyone.

And here the second part of the MySql compatibility check.

Please review and test    😃 

Tobi

### CODE REVIEW

File: lib/mshoplib/setup/TablesUpdateCharsetCollation.php [hunk 1]
```diff
@@ -171,6 +171,40 @@ protected function checkColumns( \Aimeos\MW\Setup\DBSchema\Iface $schema, $table
...
+		if ( !$schema instanceof \Aimeos\MW\Setup\DBSchema\Mysql )
+		{
+			return true;
+		}
+
+		// Ask the database for it's type and version.
+		// A MariaDB gets identified as a MySql 5.5.5 by doctrine.
+		// We can not rely on it to tell us the actual version and ask
+		// the server directly.
+		$conn = $this->acquire();
```
Comment (aimeos) [inline lib/mshoplib/setup/TablesUpdateCharsetCollation.php]: You can use getValue() instead: https://github.com/aimeos/aimeos-core/blob/master/lib/mwlib/src/MW/Setup/Task/Base.php#L197

```
$version = $this->getValue( 'SELECT version() AS "version"', 'version', 'db-customer' );

File: lib/mshoplib/setup/TablesUpdateCharsetCollation.php [hunk 2]
```diff
@@ -171,6 +171,40 @@ protected function checkColumns( \Aimeos\MW\Setup\DBSchema\Iface $schema, $table
...
+		$result = $conn->create( 'SELECT version()' )->execute();
+		// Something like '10.1.29-MariaDB' or '5.6.33-0ubuntu0...'
+		$version = $result->fetch();
+		$version = $version['version()'];
+		$result->finish();
+		$this->release($conn);
+
+		if( ( strpos( $version, 'MariaDB' ) !== false && version_compare( $version, '10.2', '>=' ) ) ||
+			version_compare( $version, '5.7', '>=' )
```
Comment (aimeos) [inline lib/mshoplib/setup/TablesUpdateCharsetCollation.php]: Do you need to check for `strpos( $version, 'MySQL' ) !== false` too?

Comment (guelzow) [reply lib/mshoplib/setup/TablesUpdateCharsetCollation.php]: '!== false' means it is somewhere in there.

The string may look like '10.1.29-MariaDB'. The distro might also add stuff at the end.

**Comment by guelzow** [comment]
What about the unit tests?

**Comment by aimeos** [comment]
The unit tests are currently written in a way that they work with utf8 and utf8mb4 at the same time

File: lib/mshoplib/setup/TablesUpdateCharsetCollation.php (hunk 2 above)
Comment (guelzow) [reply lib/mshoplib/setup/TablesUpdateCharsetCollation.php]: The MySql does not claim to be a MySql in its version number.
The dev system I'm currently working with replies with '5.6.33-0ubuntu0.14.04.1'

**Comment by guelzow** [comment]
One of them is failing.
Right now I try to find out why the shell is complaining about missing media files . . .

File: lib/mshoplib/setup/TablesUpdateCharsetCollation.php (hunk 2 above)
Comment (aimeos) [reply lib/mshoplib/setup/TablesUpdateCharsetCollation.php]: Then your check is OK :-)

**Comment by guelzow** [comment]
The setup throws a 
> No "media/path/to/folder/example1.jpg" ref ID found for "media"

trying to find media record that are not there.

The task
> Adding media test data

is not being executed.

Tracing back the list of the pre-dependencies, it all hinges on the class `CustomerListAddTestData`.
It gets asked for it's dependencies, but the tasks in the Multiple manager do not seem to be complete.

Any ideas how to solve this?

**Comment by aimeos** [comment]
There's another issue report that suggests a different solution: https://github.com/aimeos/aimeos-typo3/issues/95

**Comment by guelzow** [comment]
> The DYNAMIC row format maintains the efficiency of storing the entire row in the index node if it fits (as do the COMPACT and REDUNDANT formats), but the DYNAMIC row format avoids the problem of filling B-tree nodes with a large number of data bytes of long columns. 

Sounds interesting.
I think I'll take a look at the Barracuda file format.

**Comment by aimeos** [comment]
We've fixed the core tests but one issue in your code (tabs vs. spaces) is left

**Comment by guelzow** [comment]
Heh, sorry about the spaces . . .

File: lib/mshoplib/setup/TablesUpdateCharsetCollation.php (hunk 1 above)
Comment (guelzow) [reply lib/mshoplib/setup/TablesUpdateCharsetCollation.php]: ummm . . . 
The getValue method uses `$this->getConnection` which is marked as deprecated, so I opted for `acquire()` and `release()`.

Shall I change it?

Comment (aimeos) [reply lib/mshoplib/setup/TablesUpdateCharsetCollation.php]: That has been fixed in the master branch and it's not using getConnection() any more

**Comment by aimeos** [comment]
What about https://github.com/aimeos/aimeos-typo3/issues/95 and execute
```ALTER TABLE {here-the-table-name} ROW_FORMAT=DYNAMIC```
instead?

**Comment by guelzow** [comment]
I just talked with our admin about that.

To be able to use barracuda, you need to get the 'innodb_file_format' variable to 'Barracuda'.
The Barracuda format is standard for MySql 5.7 and MariaDB 10.2.
Both of them having no trouble with our problem (if I would have to guess, it's because of Barracuda).

The main problem is that you may not be able to change the 'innodb_file_format', because you may not have the access-rights to do so.
Please correct me if I'm wrong, because Barracuda sounds like a mayor speed upgrade.
With my tests a few days ago, I noticed that 10.2 is much faster than 10.1.

**Comment by guelzow** [comment]
You will get a:
> 15:38:54	ALTER TABLE  mshop_customer ROW_FORMAT=DYNAMIC	0 row(s) affected, 2 warning(s): 1478 InnoDB: ROW_FORMAT=DYNAMIC requires innodb_file_format > Antelope. 1478 InnoDB: assuming ROW_FORMAT=COMPACT. Records: 0  Duplicates: 0  Warnings: 2	0.328 sec

```sql
SET @@GLOBAL.innodb_file_format = 'Barracuda';
ALTER TABLE MyGuests ROW_FORMAT=DYNAMIC;
SET @@GLOBAL.innodb_file_format = 'Antelope';
```
Should work.

**Comment by aimeos** [comment]
So we have to change the SQL statements to: `ALTER TABLE "%1$s" CONVERT TO CHARACTER SET 'utf8mb4' COLLATE 'utf8mb4_bin', ROW_FORMAT = DYNAMIC` only and add `SET @@GLOBAL.innodb_file_format = 'Barracuda';` before and can skip all additional checks? MySQL 5.5+ as minimum requirement is OK as it's more than five years old :-)

**Comment by aimeos** [comment]
Doesn't work as it requires SUPER privileges not available in hosting environments:
`Access denied; you need (at least one of) the SUPER privilege(s) for this operation`

**Comment by aimeos** [comment]
Slightly improved and more documentation added:
https://github.com/aimeos/aimeos-core/commit/34e25f5928c91aebd268bca426e5a5c2a97da647

**Pull request merged**